code/modules/mob/observer/eye/blueprints_eye.dm

```diff
@@ -0,0 +1,275 @@
+
+/mob/observer/eye/blueprints/proc/update_selected_turfs(var/turf/next_selected_turf, var/list/params)
+	if(!next_selected_turf)
+		return
+
+	if(!last_selected_turf) // The player has only placed down one corner of the block.
+		last_selected_turf = next_selected_turf
+		last_selected_image.loc = last_selected_turf
+		return
+
+	if(last_selected_turf.z != next_selected_turf.z) // No multi-Z areas. Contiguity checks this as well, but this is cheaper.
+		return
+	
+	var/list/new_selection = block(last_selected_turf, next_selected_turf)
+
+	if(params["shift"])		   // Shift click to remove areas from the selection.
+		selected_turfs -= new_selection
+	else
+		selected_turfs |= new_selection
+
+	last_selected_image.loc = null // Remove the last selected turf indicator image.
+
+	check_selection_validity()
+	update_images()
+	last_selected_turf = null
+
+// Completes all the necessary checks for creating new areas, starting at the turf level before checking contiguity. 
+/mob/observer/eye/blueprints/proc/check_selection_validity()
+	. = TRUE
+	errors.Cut()
+
+	if(!LAZYLEN(selected_turfs)) // Sanity check
+		errors |= "no turfs are selected"
+		return FALSE
+
+	if(selected_turfs.len > MAX_AREA_SIZE)
+		errors |= "selection exceeds max size"
+		return FALSE
+
+	for(var/turf/T in selected_turfs)
+		var/turf_valid = check_turf_validity(T)
+		. = min(., turf_valid)
+		selected_turfs[T] = turf_valid
+	
+	if(!.) return // Skip checking contiguity if there's other errors with individual turfs.
+	. = check_contiguity()
+
+/mob/observer/eye/blueprints/proc/check_turf_validity(var/turf/T)
+	. = TRUE
+	if(!T)
+		return FALSE
+	if(!(T.z in valid_z_levels))
+		errors |= "selection isn't marked on the blueprints"
+		. = FALSE
+	var/area/A = T.loc
+	if(!A) // Safety check
+		errors |= "selection overlaps unknown location"
+		return FALSE
+	if(!(A.area_flags & AREA_FLAG_IS_BACKGROUND)) // Cannot create new areas over old ones.
+		errors |= "selection overlaps other area"
+		. = FALSE
+	if(istype(T, (A.base_turf ? A.base_turf : /turf/space)))
+		errors |= "selection is exposed to the outside"
+		. = FALSE
+
+/mob/observer/eye/blueprints/proc/check_contiguity()
+	var/turf/start_turf = pick(selected_turfs)
+	var/list/pending_turfs = list(start_turf)
+	var/list/checked_turfs = list()
+	
+	while(pending_turfs.len)
+		if(LAZYLEN(checked_turfs) > MAX_AREA_SIZE)
+			errors |= "selection exceeds max size"
+			break
+		var/turf/T = pending_turfs[1]
+		pending_turfs -= T
+		for(var/dir in GLOB.cardinal)	// Floodfill to find all turfs contiguous with the randomly chosen start_turf.
+			var/turf/NT = get_step(T, dir)
+			if(!isturf(NT) || !(NT in selected_turfs) || (NT in pending_turfs) || (NT in checked_turfs))
+				continue
+			pending_turfs += NT	
+
+		checked_turfs += T
+	
+	var/list/incontiguous_turfs = (selected_turfs.Copy() - checked_turfs) 
+
+	if(LAZYLEN(incontiguous_turfs)) // If turfs still remain in incontiguous_turfs, there are non-contiguous turfs in the selection.
```

Comment:
Non-blocking nitpick, Google reckons `incontiguous` isn't a word and recommends `noncontiguous`, @Techhead0 recommends `discontiguous`.